src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala

```diff
@@ -322,6 +326,24 @@ final class TrialService
     )
   }
 
+  private def finalizeUser(userInfo: UserInfo): Future[PerRequestMessage] = {
+    import TrialStates._
+
+    // Get user's trial status, check and update the current state if it's a valid transition
+    // NB: We are being lenient and are not complaining when a user was already terminated previously
+    thurloeDao.getTrialStatus(userInfo.id, userInfo) flatMap { status =>
+      if (Finalized.isAllowedFrom(status.state)) {
+        thurloeDao.saveTrialStatus (userInfo.id, userInfo, status.copy (state = Some (Finalized) ) ) flatMap {
+          case Success (_) => Future (RequestComplete (NoContent) )
+          case Failure (ex) => Future (RequestComplete (InternalServerError, ex.getMessage) )
+        }
```

Comment:
In the case where a finalized user attempts to re-finalize, this code sends a write to Thurloe. We don't need to contact Thurloe at all, and shouldn't write when we don't have to ... if a finalized user attempts to re-finalize, it should be a complete noop.

Reply:
Totally! To fix it, I ended up introducing a nested if block, which looks ugly. If you have suggestions on how to make it more elegant and find it worthwhile, please let me know.

Reply:
Can you do
```
match (Finalized.isAllowedFrom(state), state.contains(Terminated)) {
  case (true, true) => ...
...
}
```

Reply:
Done. Thanks!